python/dgl/distributed/dist_graph.py

```diff
@@ -0,0 +1,506 @@
+        shape = [s for s in shape]
+        shape[0] = g.number_of_nodes()
+        self._shape = tuple(shape)
+        self._dtype = dtype
+
+    def __getitem__(self, idx):
+        return self.kvstore.pull(name=self.name, id_tensor=idx)
+
+    def __setitem__(self, idx, val):
+        # TODO how do we want to support broadcast.
+        self.kvstore.push(name=self.name, id_tensor=idx, data_tensor=val)
+
+    def __len__(self):
+        return self._shape[0]
+
+    @property
+    def shape(self):
+        ''' Return the shape of the distributed tensor. '''
+        return self._shape
+
+    @property
+    def dtype(self):
+        ''' Return the data type of the distributed tensor. '''
+        return self._dtype
+
+
+class NodeDataView(MutableMapping):
+    """The data view class when G.nodes[...].data is called.
+
+    See Also
+    --------
+    dgl.DGLGraph.nodes
+    """
+    __slots__ = ['_graph', '_graph_name']
+
+    def __init__(self, g, graph_name):
+        self._graph = g
+        self._graph_name = graph_name
+
+    def __getitem__(self, key):
+        return DistTensor(self._graph, _get_ndata_name(key))
+
+    def __setitem__(self, key, val):
+        raise DGLError("DGL doesn't support set items. Please call init_ndata.")
+
+    def __delitem__(self, key):
+        #TODO how to delete data in the kvstore.
+        pass
+
+    def __len__(self):
+        return self._graph.number_of_nodes()
+
+    def __iter__(self):
+        return iter(self._graph._get_all_ndata_names())
+
+    def __repr__(self):
+        name_list = self._graph._get_all_ndata_names()
+        reprs = {}
+        for name in name_list:
+            dtype, shape, _ = self._graph._client.get_data_meta(_get_ndata_name(name))
+            reprs[name] = 'DistTensor(shape={}, dtype={})'.format(str(shape), str(dtype))
+        return repr(reprs)
+
+class EdgeDataView(MutableMapping):
+    """The data view class when G.edges[...].data is called.
+
+    See Also
+    --------
+    dgl.DGLGraph.edges
+    """
+    __slots__ = ['_graph', '_graph_name']
+
+    def __init__(self, graph, graph_name):
+        self._graph = graph
+        self._graph_name = graph_name
+
+    def __getitem__(self, key):
+        return DistTensor(self._graph, _get_edata_name(key))
+
+    def __setitem__(self, key, val):
+        raise DGLError("DGL doesn't support set items. Please call init_ndata.")
+
+    def __delitem__(self, key):
+        #TODO
+        pass
+
+    def __len__(self):
+        return self._graph.number_of_edges()
+
+    def __iter__(self):
+        return iter(self._graph._get_all_edata_names())
+
+    def __repr__(self):
+        name_list = self._graph._get_all_edata_names()
+        reprs = {}
+        for name in name_list:
+            dtype, shape, _ = self._graph._client.get_data_meta(_get_edata_name(name))
+            reprs[name] = 'DistTensor(shape={}, dtype={})'.format(str(shape), str(dtype))
+        return repr(reprs)
+
+
+class DistGraphServer(KVServer):
```

Comment:
I'm not sure if this is the correct inheritance. Suppose I have two tensors: an "h" node feature tensor and a "w" edge feature tensor. Are they stored in the same KVStore service? Does KVStore support storing distributed tensors that are partitioned differently?